### PR TITLE
Erigon 3: Prune canonical markers - `CanonicalHash`, `HeaderNumber` from chaindata

### PR DESCRIPTION
**Pull request by Giulio2002**
We prune `CanonicalHash` and `HeaderNumber ` in `chaindata`. this leads to a -99% in these 2 tables and reduce their total size to 40 pages in TOTAL (from >500k pages).

Benchmarks:

```
Total time taken to query blocks from 1000000 to 3000000: 1720.1410 seconds (OLD)
Total time taken to query blocks from 1000000 to 3000000: 1747.4729 seconds (New)

Amortized delta cost: 0.1ms/req
```

Is the RPC working for those ranges? Yes.

## How does it work?

Query canonical hash from snapshots by: `viewHeaderByNumber` -> `header.Hash()` if db hit is missed
Query header number from snapshots by: `viewHeaderByHash` -> `header.Num64` if db hit is missed

### CODE REVIEW

File: turbo/snapshotsync/freezeblocks/block_reader.go [hunk 1]
```diff
@@ -458,17 +500,13 @@ func (r *BlockReader) HeaderByHash(ctx context.Context, tx kv.Getter, hash commo
 var emptyHash = common.Hash{}
 
 func (r *BlockReader) CanonicalHash(ctx context.Context, tx kv.Getter, blockHeight uint64) (h common.Hash, err error) {
-	h, err = rawdb.ReadCanonicalHash(tx, blockHeight)
-	if err != nil {
-		return h, err
+	// Handle nil case with db hit
+	if r == nil {
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/freezeblocks/block_reader.go:504]: don't need this `if`

File: turbo/snapshotsync/freezeblocks/block_reader.go [hunk 2]
```diff
@@ -458,17 +500,13 @@ func (r *BlockReader) HeaderByHash(ctx context.Context, tx kv.Getter, hash commo
...
+		return rawdb.ReadCanonicalHash(tx, blockHeight) // Fallback to db if not found in snapshots
 	}
-	if h != emptyHash {
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/freezeblocks/block_reader.go:465]: need this check

File: turbo/snapshotsync/freezeblocks/block_reader.go [hunk 3]
```diff
@@ -458,17 +500,13 @@ func (r *BlockReader) HeaderByHash(ctx context.Context, tx kv.Getter, hash commo
...
-		return h, nil
-	}
-
 	seg, ok, release := r.sn.ViewSingleFile(coresnaptype.Headers, blockHeight)
 	if !ok {
-		return
+		return rawdb.ReadCanonicalHash(tx, blockHeight) // Fallback to db if not found in snapshots
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/freezeblocks/block_reader.go:509]: for performance must be opposite: 
- check db first
- if not in db read from file (it's more expensive)

File: turbo/snapshotsync/freezeblocks/block_reader.go [hunk 4]
```diff
@@ -379,7 +397,7 @@ func (r *BlockReader) HeaderByNumber(ctx context.Context, tx kv.Getter, blockHei
 	}
 
 	if tx != nil {
-		blockHash, err := rawdb.ReadCanonicalHash(tx, blockHeight)
+		blockHash, err := r.CanonicalHash(ctx, tx, blockHeight)
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/freezeblocks/block_reader.go:400]: seems here can use `rawdb.ReadCanonicalHash`

File: turbo/snapshotsync/freezeblocks/block_reader.go [hunk 5]
```diff
@@ -231,6 +240,13 @@ func (r *RemoteBlockReader) Body(ctx context.Context, tx kv.Getter, hash common.
 	}
 	return block.Body(), uint32(len(block.Body().Transactions)), nil
 }
+func (r *RemoteBlockReader) IsCanonical(ctx context.Context, tx kv.Getter, hash common.Hash, blockHeight uint64) (bool, error) {
+	expected, err := rawdb.ReadCanonicalHash(tx, blockHeight)
```
Comment (AskAlexSharov) [inline turbo/snapshotsync/freezeblocks/block_reader.go:244]: not enough

Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/block_reader.go:244]: can you elaborate?

File: turbo/snapshotsync/freezeblocks/block_reader.go (hunk 4 above)
Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/block_reader.go:400]: mh, I prefer using `r.CanonicalHash`, whenever I can, I just should regardless. but yes I see your point, but this is just to be safe and consistent

File: turbo/snapshotsync/freezeblocks/block_reader.go (hunk 1 above)
Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/block_reader.go:504]: ok done

File: turbo/snapshotsync/freezeblocks/block_reader.go (hunk 2 above)
Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/block_reader.go:465]: done

File: turbo/snapshotsync/freezeblocks/block_reader.go (hunk 3 above)
Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/block_reader.go:509]: done

**Comment by Giulio2002** [comment]
+ https://github.com/erigontech/interfaces/pull/227

File: turbo/snapshotsync/freezeblocks/block_reader.go (hunk 5 above)
Comment (Giulio2002) [reply turbo/snapshotsync/freezeblocks/block_reader.go]: oh i see... this got fixed

**Review by AskAlexSharov — APPROVED** [review]

**Pull request merged**